### PR TITLE
Implement specialized pool support in the scheduler

### PR DESCRIPTION
**Pull request by mihails-strasuns-sociomantic**
Shows planned design and passes most simple test. PR right now only exists as form of preview and code coverage report abuse.

TODO:

- [x] More comprehensive tests
- [x] Review / fix documentation
- [ ] Test using one of existing applications extensively using tasks

### CODE REVIEW

File: src/ocean/task/internal/FiberPoolEager.d [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+        Params:
+            task = derivative from `ocean.task.Task` defining some application
+                task to execute
+
+    ***************************************************************************/
+
+    public void run ( Task task )
+    {
+        auto fiber = this.get();
```
Comment (gavin-norman-sociomantic) [inline src/ocean/task/internal/FiberPoolEager.d:62]: So this is an unlimited pool of fibers, right?

File: src/ocean/task/internal/SpecializedPools.d [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    }
+
+    /***************************************************************************
+
+        Runs task in one of dedicated worker fiber pools if it is registered in
+        the mapping. Returns 'false' otherwise.
+
+    ***************************************************************************/
+
+    bool run ( Task task )
```
Comment (gavin-norman-sociomantic) [inline src/ocean/task/internal/SpecializedPools.d]: Do you anticipate use cases where the caller isn't sure whether the task is handled by this class or not? I would have expected it to throw if the task isn't of a supported type.

**Review by gavin-norman-sociomantic — COMMENTED** [review]
Interesting that all specialized task pools are unlimited. What's the concept behind that?

File: src/ocean/task/internal/FiberPoolEager.d (hunk 1 above)
Comment (mihails-strasuns-sociomantic) [reply src/ocean/task/internal/FiberPoolEager.d:62]: Yes. My line of though went like that - we are only expected to use those pools for real-time kind of handling where waiting in queue is not appropriate. Like HTTP connection handler. If we allow the limit, what would be a reasonable fallback for case when it is reached? Close the connection, bring down the app? Nothing really made sense. It feels like having unbound limit and just monitoring graphs is the only practical approach.

File: src/ocean/task/internal/SpecializedPools.d (hunk 2 above)
Comment (mihails-strasuns-sociomantic) [reply src/ocean/task/internal/SpecializedPools.d]: Right now it is used to simplify code in scheduler. It does something like `if (!custom_pools.run(task)) generic_pools.run(task);`.

Comment (gavin-norman-sociomantic) [reply src/ocean/task/internal/SpecializedPools.d]: Fair enough.

File: src/ocean/task/internal/FiberPoolEager.d (hunk 1 above)
Comment (gavin-norman-sociomantic) [reply src/ocean/task/internal/FiberPoolEager.d:62]: I see. This works fine for normal usage where the number of connections (say) gradually grows over time. We also need to consider erroneous or malicious behaviour, though, where the number of connections could grow extremely quickly. OOM is not an acceptable solution ;)

**Comment by mihails-strasuns-sociomantic** [comment]
I think this is ready for review. Changes from last update:

- new higher level test which I used to manually trace scheduling logic
- bunch of small fixes everywhere

Testing with some real application would be nice but I don't think it is important enough right now to distract app maintainers (as the effort is non-trivial).

File: src/ocean/task/Scheduler.d [hunk 3]
```diff
@@ -27,7 +27,6 @@ import ocean.transition;
 import ocean.core.Enforce;
 import ocean.core.Verify;
 import ocean.core.Traits;
-import ocean.core.array.Mutation : reverse;
```
Comment (gavin-norman-sociomantic) [inline src/ocean/task/Scheduler.d:30]: Commit message comment:
"This is the inherent part of running task, not matter the context it is being run from."
->
"This is an inherent part of running a task, not a matter for the context it is being run from."
(Assuming I understand what you want to say.)

File: src/ocean/task/Task.d [hunk 4]
```diff
@@ -456,8 +456,39 @@ public abstract class Task : ISuspendable
         debug_trace("<{}> start of main function", cast(void*) this);
 
         scope(exit)
+        {
             this.state_bitmask |= TaskState.Finished;
 
+            if (this.termination_hooks.length)
```
Comment (gavin-norman-sociomantic) [inline src/ocean/task/Task.d:473]: This seems like a potentially dangerous change: running the termination hooks inside the task, rather than from the calling context.

File: src/ocean/task/internal/FiberPoolWithQueue.d [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+    ***************************************************************************/
+
+    public TaskQueueFullCB task_queue_full_cb;
+
+    /**************************************************************************
+
+        Constructor
+
+        Params:
+            stack_size = fiber stack size to use in this poll
```
Comment (gavin-norman-sociomantic) [inline src/ocean/task/internal/FiberPoolWithQueue.d]: Incomplete params doc.

File: src/ocean/task/internal/FiberPoolWithQueue.d [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+        {
+            this.queue(task);
+            return;
+        }
+
+        auto fiber = this.get();
+        debug_trace("running task <{}> via worker fiber <{}>",
+            cast(void*) task, cast(void*) fiber);
+        task.assignTo(fiber);
+        // trick task by using different method for real entry point:
```
Comment (gavin-norman-sociomantic) [inline src/ocean/task/internal/FiberPoolWithQueue.d]: This comment is a little mysterious...

File: src/ocean/task/internal/FiberPoolEager.d (hunk 1 above)
Comment (gavin-norman-sociomantic) [reply src/ocean/task/internal/FiberPoolEager.d:62]: As far as I remember from our discussion on this before, the conclusion was that this is fine/irrelevant, as the select listener already implements a limit on the number of handled connections. (And rejects any incoming connections beyond the limit.)

File: src/ocean/task/Scheduler.d [hunk 7]
```diff
@@ -69,6 +70,16 @@ struct SchedulerConfiguration
...
     /// `theScheduler.processEvents` in between scheduler dispatch cycles
     size_t suspended_task_limit = 16;
+
+    /// optional array that defines specialized worker fiber pools to be
+    /// used for handling specific task kinds. Scheduled task is checked
+    /// against this array every time thus it is not recommended to configure
+    /// it to more than a few dedicated extra pools
+    PoolDescription[] specialized_pools;
+
+    /// specialized worker fiber pool description. Tasks handled by such / pools
```
Comment (gavin-norman-sociomantic) [inline src/ocean/task/Scheduler.d]: The "/" at the end of this line seems erroneous.

File: src/ocean/task/Task.d [hunk 8]
```diff
@@ -83,6 +83,17 @@ public class WorkerFiber : core.thread.Fiber
 
     /***************************************************************************
 
+        Terminates itself by killing underlying task
```
Comment (gavin-norman-sociomantic) [inline src/ocean/task/Task.d:86]: This seems like it should be a separate commit.

File: src/ocean/task/internal/SpecializedPools.d [hunk 9]
```diff
@@ -0,0 +1,124 @@
+/*******************************************************************************
+
```
Comment (gavin-norman-sociomantic) [inline src/ocean/task/internal/SpecializedPools.d:2]: Module description missing.

File: src/ocean/task/internal/SpecializedPools.d [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+    Defines all mappings between task kinds and dedicated worker fiber pools
+    processing those.
+
+*******************************************************************************/
+
+public class SpecializedPools
+{
+    /// mapping as a simple array (expected to contain few elements)
+    private SpecializedPool[] mapping;
```
Comment (gavin-norman-sociomantic) [inline src/ocean/task/internal/SpecializedPools.d:53]: What's the reason you're not using an AA?

File: src/ocean/task/internal/SpecializedPools.d [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+    private static struct SpecializedPool
+    {
+        ClassInfo task_kind;
+        FiberPoolEager pool;
+    }
+
+    /***************************************************************************
+
+        Constructor
+
```
Comment (gavin-norman-sociomantic) [inline src/ocean/task/internal/SpecializedPools.d:65]: Params

File: src/ocean/task/internal/SpecializedPools.d [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+    {
+        foreach (description; config)
+        {
+            verify(description.task_kind !is null);
+            verify(
+                description.stack_size >= 1024,
+                "Configured stack size is suspiciously low"
+            );
+
+            this.mapping ~= SpecializedPool(
```
Comment (gavin-norman-sociomantic) [inline src/ocean/task/internal/SpecializedPools.d]: Using an AA would also allow us to detect multiple mappings for the same task type.

File: src/ocean/task/internal/SpecializedPools.d [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+                )
+            );
+        }
+    }
+
+    /***************************************************************************
+
+        Runs task in one of dedicated worker fiber pools if it is registered in
+        the mapping. Returns 'false' otherwise.
+
```
Comment (gavin-norman-sociomantic) [inline src/ocean/task/internal/SpecializedPools.d]: Params / returns.

File: src/ocean/task/internal/SpecializedPools.d [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+        return false;
+    }
+
+    /***************************************************************************
+
+        Kills all worker fibers in all pools
+
+    ***************************************************************************/
+
+    void kill ( )
```
Comment (gavin-norman-sociomantic) [inline src/ocean/task/internal/SpecializedPools.d]: Protection attribute missing on all methods. (All public, I suppose.)

File: src/ocean/task/internal/SpecializedPools.d [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+        foreach (meta; this.mapping)
+        {
+            scope iterator = meta.pool.new BusyItemsIterator;
+            foreach (ref fiber; iterator)
+                fiber.kill();
+        }
+    }
+}
+
+private void debug_trace ( T... ) ( cstring format, T args )
```
Comment (gavin-norman-sociomantic) [inline src/ocean/task/internal/SpecializedPools.d:171]: Unused?

**Review by gavin-norman-sociomantic — COMMENTED** [review]
Looking good!

**Comment by gavin-norman-sociomantic** [comment]
Oh, needs release notes, too.

File: src/ocean/task/Task.d (hunk 4 above)
Comment (mihails-strasuns-sociomantic) [reply src/ocean/task/Task.d:473]: There is no change, it already works like that (the fiber is the same).

Comment (gavin-norman-sociomantic) [reply src/ocean/task/Task.d:473]: Ah I see, `runTask` is called from inside the fiber. :+1:

File: src/ocean/task/internal/FiberPoolEager.d (hunk 1 above)
Comment (mihails-strasuns-sociomantic) [reply src/ocean/task/internal/FiberPoolEager.d:62]: Correct. Consensus was to think about implementing a limit once we have real app use case that may require it.

File: src/ocean/task/internal/SpecializedPools.d (hunk 10 above)
Comment (mihails-strasuns-sociomantic) [reply src/ocean/task/internal/SpecializedPools.d:53]: Cons: slower, inherent heap fragmentation, have to keep in mind all const-related AA bugs
Pros: nice syntax

I think for purely private entity in stdlib choice is obvious :P

**Comment by mihails-strasuns-sociomantic** [comment]
Updated: https://gist.github.com/mihails-strasuns-sociomantic/ce7257baa7584449dce0c21be29d0a7e

**Comment by mihails-strasuns-sociomantic** [comment]
I am also bumping priority because investigation on https://github.com/sociomantic-tsunami/ocean/pull/335 shows there are pretty big changes that need to be done on scheduler - and I'd prefer to do those on top of refactoring from this PR.

**Comment by mihails-strasuns-sociomantic** [comment]
Added release notes

**Comment by mihails-strasuns-sociomantic** [comment]
Green

File: src/ocean/task/internal/FiberPoolWithQueue.d (hunk 5 above)
Comment (gavin-norman-sociomantic) [reply src/ocean/task/internal/FiberPoolWithQueue.d]: Still incomplete.

File: relnotes/fiber-pool.feature.md [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+
+  initScheduler(config);
+
+  // will be processed immediately in own worker fiber pool, skipping the queue:
+  theScheduler.schedule(new MyTask);
+  ```
+
+  Such functionality is intended to be used when there are some task types that
+  need real-time handling and thus can't afford waiting in the queue. It can
+  also be used for case when some tasks are very long-living would hold worker
```
Comment (gavin-norman-sociomantic) [inline relnotes/fiber-pool.feature.md]: Grammar / clarity: "It can also be used for the case where some very long-lived tasks would occupy worker fibers from the shared pool forever, if processed normally."

File: src/ocean/task/internal/FiberPoolWithQueue.d (hunk 5 above)
Comment (mihails-strasuns-sociomantic) [reply src/ocean/task/internal/FiberPoolWithQueue.d]: https://github.com/sociomantic-tsunami/ocean/pull/293/files#diff-2c0e70ae37423c24132186053ded1353R68

Comment (gavin-norman-sociomantic) [reply src/ocean/task/internal/FiberPoolWithQueue.d]: Ah, the comment fix was in the wrong commit.

**Review by gavin-norman-sociomantic — COMMENTED** [review]
A couple of minor comments.

**Comment by mihails-strasuns-sociomantic** [comment]
Updated:
- relnotes wording
- move doc fix to proper commit

**Review by gavin-norman-sociomantic — APPROVED** [review]

**Comment by gavin-norman-sociomantic** [comment]
Feel free to auto-merge when green.

File: src/ocean/task/internal/FiberPoolWithQueue.d [hunk 17]
```diff
@@ -0,0 +1,249 @@
...
+            stack_size = fiber stack size to use in this poll
+            limit = limit to pool size. If set to 0 (default), there is no
+                app limit and pool growth will be limited only by OS
+                resources
+            process_events = reference to scheduler method which does temporary
+                suspend, passed to constructor to avoid circular dependency
+
+    **************************************************************************/
+
+    this ( size_t queue_limit, size_t stack_size, size_t limit,
```
Comment (daniel-zullo) [inline src/ocean/task/internal/FiberPoolWithQueue.d:70]: missing protection attribute

File: src/ocean/task/internal/FiberPoolEager.d [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+    /**************************************************************************
+
+        Constructor
+
+        Params:
+            stack_size = fiber stack size to use in this poll
+
+    **************************************************************************/
+
+    this ( size_t stack_size )
```
Comment (daniel-zullo) [inline src/ocean/task/internal/FiberPoolEager.d:42]: missing protection attribute

File: src/ocean/task/internal/FiberPoolEager.d [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+import ocean.meta.types.Qualifiers;
+import ocean.core.Enforce;
+import ocean.task.internal.FiberPool;
+import ocean.task.Task;
+
+debug (TaskScheduler)
+    import ocean.io.Stdout;
+
+/// Ditto
+class FiberPoolEager : FiberPool
```
Comment (daniel-zullo) [inline src/ocean/task/internal/FiberPoolEager.d:31]: missing protection attribute

File: src/ocean/task/Scheduler.d [hunk 20]
```diff
@@ -299,12 +318,15 @@ final class Scheduler
 
     public void shutdown ( )
     {
-        debug_trace("Shutting down initiated. {} queued tasks will be " ~
-            " discardead, {} suspended tasks will be killed",
-            this.queued_tasks.length(), this.suspended_tasks.length());
+        debug_trace(
+            "Shutting down initiated. {} queued tasks will be " ~
+                " discardead, {} suspended tasks will be killed",
```
Comment (daniel-zullo) [inline src/ocean/task/Scheduler.d:323]: typo *discardead

File: src/ocean/task/Task.d [hunk 21]
```diff
@@ -299,8 +310,8 @@ public abstract class Task : ISuspendable
     public void suspend ( )
     {
         assert (this.fiber !is null);
```
Comment (daniel-zullo) [inline src/ocean/task/Task.d:312]: shouldn't it be replaced to `verify()` too?

File: src/ocean/task/internal/FiberPoolEager.d [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+        3-Clause BSD License (see LICENSE_BSD.txt for details).
+
+*******************************************************************************/
+
+module ocean.task.internal.FiberPoolEager;
+
+import core.thread : Fiber;
+
+import ocean.meta.types.Qualifiers;
+import ocean.core.Enforce;
```
Comment (daniel-zullo) [inline src/ocean/task/internal/FiberPoolEager.d:23]: this import should go on the top of ocean imports

File: src/ocean/task/internal/FiberPoolEager.d [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+
+/// Ditto
+class FiberPoolEager : FiberPool
+{
+    /**************************************************************************
+
+        Constructor
+
+        Params:
+            stack_size = fiber stack size to use in this poll
```
Comment (daniel-zullo) [inline src/ocean/task/internal/FiberPoolEager.d:38]: s/poll/pool ?

File: src/ocean/task/internal/FiberPoolWithQueue.d [hunk 24]
```diff
@@ -0,0 +1,249 @@
...
+    public TaskQueueFullCB task_queue_full_cb;
+
+    /**************************************************************************
+
+        Constructor
+
+        Params:
+            queue_limit = max size of task queue, recommened to be at least as
+                large as pool limit
+            stack_size = fiber stack size to use in this poll
```
Comment (daniel-zullo) [inline src/ocean/task/internal/FiberPoolWithQueue.d:61]: ditto *poll

File: src/ocean/task/internal/SpecializedPools.d [hunk 25]
```diff
@@ -0,0 +1,178 @@
...
+*******************************************************************************/
+
+module ocean.task.internal.SpecializedPools;
+
+import ocean.meta.types.Qualifiers;
+import ocean.task.internal.FiberPoolEager;
+import ocean.task.Task;
+import ocean.core.Verify;
+import ocean.core.Enforce;
+import ocean.core.Optional;
```
Comment (daniel-zullo) [inline src/ocean/task/internal/SpecializedPools.d:22]: Should the imports be sorted?

**Review by daniel-zullo — COMMENTED** [review]
A bit late for my minor comments but I'd submit them anyway

File: src/ocean/task/Task.d (hunk 21 above)
Comment (mihails-strasuns-sociomantic) [reply src/ocean/task/Task.d:312]: No, `task` package is currently the only one which uses both as intended (depending on what kind of condition is protected against), contrary to rest of ocean which was converted 1-to-1.

I will likely have a full re-read of all scheduler source code soon and will also re-evaluate all assert/verify usage cases.

File: src/ocean/task/internal/SpecializedPools.d (hunk 25 above)
Comment (mihails-strasuns-sociomantic) [reply src/ocean/task/internal/SpecializedPools.d:22]: No, there is no particular import order requirement for our libs/apps.

**Comment by mihails-strasuns-sociomantic** [comment]
@daniel-zullo-sociomantic thanks, I will submit separate style cleanup PR

File: src/ocean/task/Task.d (hunk 21 above)
Comment (daniel-zullo) [reply src/ocean/task/Task.d:312]: ok, good to know that

File: src/ocean/task/internal/SpecializedPools.d (hunk 25 above)
Comment (daniel-zullo) [reply src/ocean/task/internal/SpecializedPools.d:22]: fair enough
